src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,494 @@
+                                    actualTypeArgument.toErasure().getName());
+                            });
+                    }
+                }
+            }
+            return isImmutable;
+        }
+
+        @Override
+        public void check(JavaField item, ConditionEvents events) {
+            if (!isImmutable(item)) {
+                final String message = String
+                    .format(Locale.ROOT, "Field <%s> should %s in %s",
+                            item.getFullName(), getDescription(),
+                            item.getSourceCodeLocation().toString());
+                events.add(SimpleConditionEvent.violated(item, message));
+            }
+        }
+    }
+
+    /**
+     * DescribedPredicate defining condition for a field to be a module property.
+     */
+    private static final class ModulePropertyPredicate extends DescribedPredicate<JavaField> {
+
+        private ModulePropertyPredicate() {
+            super("module properties");
+        }
+
+        /**
+         * Whether a field is a module property or not.
+         *
+         * @param javaField field to check
+         * @return {@code true} if field is a module property
+         */
+        private static boolean isModuleProperty(JavaField javaField) {
+            boolean result = false;
+            final JavaClass containingClass = javaField.getOwner();
+            final ModuleDetails moduleDetails = MODULE_DETAILS_MAP.get(
+                containingClass.getFullName());
+            if (moduleDetails != null) {
+                final List<ModulePropertyDetails> properties = moduleDetails.getProperties();
+                result = properties.stream()
+                    .map(ModulePropertyDetails::getName)
+                    .anyMatch(moduleName -> moduleName.equals(javaField.getName()));
+            }
+            return result;
+        }
+
+        @Override
+        public boolean apply(JavaField input) {
+            return isModuleProperty(input);
+        }
+    }
+
+    /**
+     * DescribedPredicate defining condition for a class to have immutable fields.
+     */
+    private static final class ImmutableFieldsPredicate extends DescribedPredicate<JavaClass> {
+        private ImmutableFieldsPredicate() {
+            super("immutable fields");
+        }
+
+        @Override
+        public boolean apply(JavaClass input) {
+            final Set<JavaField> fields = input.getFields();
+            return fields.stream()
+                .filter(javaField -> {
+                    return !ModulePropertyPredicate.isModuleProperty(javaField)
+                        && !SUPPRESSED_FIELDS_IN_MODULES.contains(javaField.getFullName());
+                })
+                .allMatch(javaField -> {
+                    final Set<JavaModifier> javaFieldModifiers = javaField.getModifiers();
+                    boolean result = javaFieldModifiers.contains(JavaModifier.FINAL)
```

Comment:
`final` does not make a field immutable. An array list can be made final but you can still add new items to it.

Reply:
Yes, it doesn't for an array list, but it gives reference immutability for the field (not for data inside the field), so it will help in partially assuring immutability.

Reply:
It won't if this is all connected to Stateless. Reference immutability won't help when multiple threads are rummaging through this check. If the field is mutable (public or private), multiple threads can come in and read and write to the field at the same time. A field must be completely immutable (regardless of modifiers) or we should throw an violation on it.

Reply:
Ok, I think I got lost somewhere on immutability and not needing `final`. It does need `final` otherwise the value can be swapped out. Let me know if I led you astray in any other comments about this. I still stand by private/public has no connection to immutability. `MSG_` is still immutable since it is a final String.

Reply:
Done, removed public, private part.